src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -290,45 +290,27 @@ internal bool TryWriteDataExtensionProperty(Utf8JsonWriter writer, T value, Json
             bool success;
             JsonDictionaryConverter<T> dictionaryConverter = (JsonDictionaryConverter<T>)this;
 
-            if (ClassType == ClassType.Value)
```

Comment:
I am not sure if it is OK to remove this (it's possible that it is OK to do, but is surprising to me). What happens when the user defines and register their own `JsonConverter<T>` for dictionaries?

@steveharter, @layomia - please review this when you get a chance. I suspect this branch was needed for something (which means we need to provide the test case).

Reply:
I touched on this in Note 1 in the above comments, but I'm not sure it's possible.  The line immediately above this one casts the current instance to a `JsonDictionaryConverter<T>` which is marked `internal`, so my assumption was users wouldn't be able to inherit it (unless there's something I'm not thinking of or aware of?)

I can certainly leave this in there if that's the call, but I think then testing and getting coverage on these lines would be difficult without introducing an `InternalsVisibleTo` relationship (or similar internals-testing-pattern) between the src and test project.  That or adding an inherited class in the src project for the purposes of testing.

Reply:
There is a bug with writing extension data if there is a custom converter for the extension property: https://github.com/dotnet/runtime/issues/32903. This is the reason why the L291 is not problematic (`InvalidCastException`) - we throw an NRE up stack.

We'll likely need this branch when we fix that bug, so for this PR we should revert this change.

Reply:
reverted in 256de92